src/backend/vulkan/src/result.rs

```diff
@@ -0,0 +1,91 @@
+
+use ash::vk;
+
+use hal::adapter::DeviceCreationError;
+
+// Generic error codes from Vulkan
+#[derive(Debug)]
+pub(crate) enum Error {
+    OutOfHostMemory,
+    OutOfDeviceMemory,
+    InitializationFailed,
+    DeviceLost,
+    MemoryMapFailed,
+    LayerNotPresent,
+    ExtensionNotPresent,
+    FeatureNotPresent,
+    IncompatibleDriver,
+    TooManyObjects,
+    FormatNotSupported,
+    FragmentedPool ,
+    SurfaceLostKhr,
+    NativeWindowInUseKhr,
+    OutOfDateKhr,
+    IncompatibleDisplayKhr,
+    ValidationFailedExt,
+    // Not an actual vulkan error, but handle the case where an implementation
+    // might return an unkown error.
+    Unknown,
+}
+
+impl From<vk::Result> for Error {
+    fn from(result: vk::Result) -> Self {
+        use ash::vk::Result::*;
+        match result {
+            ErrorOutOfHostMemory => Error::OutOfHostMemory,
+            ErrorOutOfDeviceMemory => Error::OutOfDeviceMemory,
+            ErrorInitializationFailed => Error::InitializationFailed,
+            ErrorDeviceLost => Error::DeviceLost,
+            ErrorMemoryMapFailed => Error::MemoryMapFailed,
+            ErrorLayerNotPresent => Error::LayerNotPresent,
+            ErrorExtensionNotPresent => Error::ExtensionNotPresent,
+            ErrorFeatureNotPresent => Error::FeatureNotPresent,
+            ErrorIncompatibleDriver => Error::IncompatibleDriver,
+            ErrorTooManyObjects => Error::TooManyObjects,
+            ErrorFormatNotSupported => Error::FormatNotSupported,
+            ErrorFragmentedPool => Error::FragmentedPool,
+            ErrorSurfaceLostKhr => Error::SurfaceLostKhr,
+            ErrorNativeWindowInUseKhr => Error::NativeWindowInUseKhr,
+            ErrorOutOfDateKhr => Error::OutOfDateKhr,
+            ErrorIncompatibleDisplayKhr => Error::IncompatibleDisplayKhr,
+            ErrorValidationFailedExt => Error::ValidationFailedExt,
+            _ => Error::Unknown,
+        }
+    }
+}
+
+// Impl `From<Error>` for various HAL error types.
+//
+// Syntax:
+//    #HalError {
+//       #VulkanError => #HalErrorVariant,
+//    }
+macro_rules! from_error {
+    { $($name:ident { $($base_error:ident => $err:ident,)* },)* } => {
+        $(
+            impl From<Error> for $name {
+                fn from(err: Error) -> Self {
+                    match err {
+                        $(
+                            Error::$base_error => $name::$err,
+                        )*
+                        _ => unreachable!("Unexpected error code ({:?}). Non specification conformant driver.", err),
```

Comment:
This part is nagging on me a bit, as I don't have experience on how much drivers do things by the spec regarding return codes..

Reply:
Agree. Can we have some sort of "unrecognized" error instead with logging the case to `error!`?

Reply:
Addind an `Unrecognized` variant to all error structs?

Reply:
Nah, on a second thought, we should be respecting the spec.